src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,178 @@
+  if (eCANComParamterAmount != parser.parseParameters()) {
+    DEVLOG_ERROR("[CAN Layer] Parsing of the parameters failed.\n");
+    return eRetVal;
+  }
+
+  SCANParameters can_params = setupCANInternals(parser);
+
+  switch (mFb->getComServiceType()) {
+  case e_Server:
+    DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                 "pattern. Please use a Publisher block to send data.");
+    eRetVal = e_InitTerminated;
+    break;
+  case e_Client:
+    DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                 "pattern. Please use a Subscribe block to receive data.");
+    eRetVal = e_InitTerminated;
+    break;
+  case e_Publisher:
+    // is handled via sendData
+    break;
+  case e_Subscriber:
+    // is handled via recvData
+    break;
+  }
+
+  DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+  CFDSelectHandler::TFileDescriptor fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+  if (CFDSelectHandler::scmInvalidFileDescriptor == fd) {
+    DEVLOG_ERROR("[CAN Layer] Socket can't be created: %s\n", strerror(errno));
+    return e_InitInvalidId;
+  }
+
+  mCANSocket = fd;
+
+  std::strncpy(mIfr.ifr_name, can_params.interfaceName.c_str(), IFNAMSIZ - 1);
+  mIfr.ifr_name[IFNAMSIZ - 1] =
+      '\0'; // Ensure that null termination is added if input is too long.
+  if (-1 == ioctl(mCANSocket, SIOCGIFINDEX, &mIfr)) {
+    DEVLOG_ERROR("[CAN Layer] Socket cannot be controlled. %s\n",
+                 strerror(errno));
+    close(mCANSocket);
+    return eRetVal;
+  }
+
+  mAddr.can_family = AF_CAN;
+  mAddr.can_ifindex = mIfr.ifr_ifindex;
+
+  if (-1 == bind(mCANSocket, (struct sockaddr *)&mAddr, sizeof(mAddr))) {
+    DEVLOG_ERROR("[CAN Layer] Binding of the socket unsuccessful: %s\n",
+                 strerror(errno));
+    close(mCANSocket);
+    return eRetVal;
+  }
+
+  this->getExtEvHandler<CFDSelectHandler>().addComCallback(mCANSocket, this);
+  DEVLOG_DEBUG("[CAN Layer] Callback handler added.\n");
+
+  mFrame.can_id = can_params.CANId;
+
+  eRetVal = e_InitOk;
+  return eRetVal;
+}
+
+CCANComLayer::~CCANComLayer() { closeConnection(); }
+
+void CCANComLayer::closeConnection() {
+  if (mCANSocket != CFDSelectHandler::scmInvalidFileDescriptor) {
+    DEVLOG_DEBUG("[CAN Layer] Closing socket\n");
+    close(mCANSocket);
+    DEVLOG_DEBUG("[CAN Layer] Unregister layer\n");
+    this->getExtEvHandler<CFDSelectHandler>().removeComCallback(mCANSocket);
+  }
+}
+
+EComResponse CCANComLayer::sendData(void *paData, unsigned int) {
+  auto *raw_data = static_cast<CIEC_ARRAY *>(
+      &(static_cast<CIEC_ANY **>(paData)[0]->unwrap()));
+
+  size_t size = raw_data->getToStringBufferSize();
+  DEVLOG_DEBUG("[CAN Layer] Size of the array is: %u.\n", size);
+  std::vector<char> buffer(size);
+
+  int nLength = raw_data->toString(buffer.data(), size);
+  if (nLength <= 0) {
+    std::memset(buffer.data(), 0, size);
+    DEVLOG_ERROR("[CAN Layer] Copying of the raw data failed.");
```

Comment:
you probably want to return here instead of keep going and try to send an invalid packet